### ISSUE TITLE
Support debug_traceTransaction

### ISSUE THREAD

**Issue by WyseNynja** [foundry-rs/foundry#1737]
### Component

Anvil

### Describe the feature you would like

I am currently working on adding support for anvil to Brownie. (https://github.com/eth-brownie/brownie/pull/1541). Brownie makes heavy use of `eth_debugTraceTransaction`

It looks like anvil already has a stub for the rpc call that returns `Not implemented`:

https://github.com/foundry-rs/foundry/blob/b8733d4e2e10dc54928cefb62015a5c26ca5f473/anvil/src/eth/api.rs#L1140-L1147

@gakonst said in telegram: "I think we should already should be able to support geth style traces. Maybe we dont expose over rpc?"

### Additional context

Brownie currently checks for `response["error"]["code"] != -32601`, so changing anvil to do that in the mean time would be helpful.

https://github.com/eth-brownie/brownie/blob/master/brownie/network/web3.py#L119-L131

**Comment by DrawnWren** [comment]
Any progress on this? I'm happy to try and pick it up over the weekend if it's stalled.

**Comment by mattsse** [comment]
this requires some effort, 

first, ethers-rs does not have support for geth tracing either, 

https://github.com/gakonst/ethers-rs/issues/558 contains (some of) the first work package, which is bindings for geth's trace types.

once we have that we need proper conversions from revm tracing to geth trace types

the current revm tracing to parity tracing conversion is handled here https://github.com/foundry-rs/foundry/blob/master/evm/src/trace/node.rs we'd need the same for geth tracing types

so to summarize

1. add geth tracing type bindings to ethers-rs
2. add tracing conversion in https://github.com/foundry-rs/foundry/blob/master/evm/src/trace/node.rs
3. convert on geth tracing RPC request, same as for parity, ref https://github.com/foundry-rs/foundry/blob/a7ea46e64c14b1f80c9b3b39fb30a03e09681ba1/anvil/src/eth/backend/mem/storage.rs#L223

**Comment by DrawnWren** [comment]
1. Submitted a PR [gakonst/ethers-rs#1469](https://github.com/gakonst/ethers-rs/pull/1469)
2 (and 3) wip.

**Comment by DrawnWren** [comment]
So, this ends up being hairier than expected. 

https://github.com/foundry-rs/foundry/blob/695bff10ab048625eb8487bd09a310b001f37383/anvil/core/src/types.rs#L13-L26

The ``tracer`` option is a string which is evaluated as JS of the form:
```
tracer = function(tx) {
   return debug.traceTransaction(tx, {tracer:
      '{' +
         'retVal: [],' +
         'step: function(log,db) {this.retVal.push(log.getPC() + ":" + log.op.toString())},' +
         'fault: function(log,db) {this.retVal.push("FAULT: " + JSON.stringify(log))},' +
         'result: function(ctx,db) {return this.retVal}' +
      '}'
   }) // return debug.traceTransaction ...
}   // tracer = function ...
```

So, it requires writing some JS built-ins on top of evaluating the JS. Do you have any preference for the JS evaluation piece? it looks like there's a library, ``js_sys`` that does JS evaluation but I'll still have to think on the built-ins.

**Comment by mattsse** [comment]
> It requires writing some JS built-ins on top of evaluating the JS. Do you have any preference for the JS evaluation piece? it looks like there's a library, js_sys that does JS evaluation but I'll still have to think on the built-ins.

sorry, hard time understanding this, why would we need to evaluate JS?

I don't know where this snippet is from.

`GethDebugTracingOptions` are the possible options the endpoint accepts.

so next step would be to convert revm traces to the new geth tracing types, right? once we have those conversions, we can deal wth the `GethDebugTracingOptions`

we implemented conversions for parity traces here https://github.com/foundry-rs/foundry/blob/master/evm/src/trace/node.rs

lmk what's unclear

**Comment by DrawnWren** [comment]
Sorry, the JS snippet is from here: https://geth.ethereum.org/docs/dapp/custom-tracer#custom-javascript-tracing

Essentially, ``debug_traceTransaction(tx_hash, options)`` where [options optionally includes a JS "tracer"](https://geth.ethereum.org/docs/rpc/ns-debug#debug_tracetransaction)

**Comment by Natanela** [comment]
+1 on this one

This could very helpful for us as for now we are not able to deeply analyze transactions that were executed on Anvil and we are forced to work side-by-side with hardhat.

**Comment by mattsse** [comment]
bumping this

**Comment by haowang0402** [comment]
@mattsse we need javascript support because `Geth` takes a javascript tracer code snippet to process the tracing results. Also, it seems that this issue is not active anymore. I would also love to see this feature, and I am down to work on this. Could you guys also assign me to this issue? Thank you!

**Comment by gakonst** [comment]
I don't feel super hot about allowing the geth-style JS tracer here FWIW. Geth itself has switched to native tracers I believe for performance reasons?

**Comment by haowang0402** [comment]
I agree that JS tracer is slower than native tracer but it's easier to integrate with existing tracer implementations out there. However, I don't have preferences between JS-style tracers and native tracers. We probably need to support the native tracers (mostly `callTracer` and `prestateTracer`) that `geth` supports.

**Comment by yhayun** [comment]
+1 on this request.
having this supported is needed for parsing internal transactions on Blocker Explorers.

**Comment by gakonst** [comment]
>  having this supported is needed for parsing internal transactions on Blocker Explorers.

got some example code?

**Comment by mattsse** [comment]
closing as `debug_traceTransaction` landed with #2868
